src/librustc_mir/borrow_check/mod.rs

```diff
@@ -703,9 +733,16 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             (sd, place_span.0),
             flow_state,
             |this, index, borrow| match (rw, borrow.kind) {
-                (Read(_), BorrowKind::Shared) => Control::Continue,
+                // Obviously an activation is compatible with its own reservation;
+                // so don't check if they interfere.
+                (Activation(_, activating), _) if index.is_reservation() &&
```

Comment:
what's the idea of this? If there's a reservation going around a loop, this could actually cause unsoundness.

Reply:
First off: Good catch. I definitely need to think more about this.

Explanation: The reason for this check here is to ensure that the activation is not treated as a conflict with its own reservation. (Any time we reach a activating use, it is by definition going to have a reservation for that borrow also in the current flow state -- because if there wasn't a reservation in the flow state, then we would not consider the use an activation.)

But your loop example is important, and may show a big flaw in how I've approached this.

I think I've made a mistake in that the current Reservations dataflow is tracking what borrows *may* reach a location, but for the scenario I'm trying to catch here, it might be necessary to use a *must*-reach semantics. Building up that flow analysis and employing it (at least for the desired filter here) might be a way to resolve the loop example you give (because with *must*-reach, a control-flow merge on entry to the loop will remove the reservations that are introduced within the loop itself).
 * Having said that, such an analysis is a pretty big addition to the framework here. I need to think more on this.

Reply:
(Another potential answer, suggested by @nikomatsakis, might be to treat the reservation in a loop as an error.)